Retry failed artifact downloads

refi64 opened the pull request:
Right now, if downloading an artifact fails due to, say, network or server errors, the runner crate simply returns an error. This isn't ideal, since *some* errors will inevitably occur; even the official runner implementation has retries for artifact downloads.

This adds a simple retry system on top of downloads: if a "retriable" error occurs (namely, reqwest-level error or a 5xx status code), then the download will be repeatedly retried up to 10 times.

In order to be able to test the change, this adds a new method to the mock API, `set_inject_artifact_download_failures`, which will cause N artifact downloads to fail before one can succeed.

Fixes: collabora/lava-gitlab-runner#102

refi64 committed:
Retry failed artifact downloads

Right now, if downloading an artifact fails due to, say, network or
server errors, the runner crate simply returns an error. This isn't
ideal, since *some* errors will inevitably occur; even the official
runner implementation has retries for artifact downloads.

This adds a simple retry system on top of downloads: if a "retriable"
error occurs (namely, reqwest-level error or a 5xx status code), then
the download will be repeatedly retried up to 10 times.

In order to be able to test the change, this adds a new method to the
mock API, `set_inject_artifact_download_failures`, which will cause N
artifact downloads to fail before one can succeed.

Fixes: collabora/lava-gitlab-runner#102

Signed-off-by: Ryan Gonzalez <ryan.gonzalez@collabora.com>

sjoerdsimons commented on gitlab-runner/src/job.rs:
```diff
@@ -63,6 +69,14 @@ pub struct Dependency<'a> {
     dependency: &'a JobDependency,
 }
 
+fn is_retriable_error(err: &ClientError) -> bool {
```
Should this be directly on `Error`

refi64 replied:
Maaaybe in the future? But right now this logic is pretty specific to artifact downloads, so we might have a slightly different standard for "retriable" elsewhere.

sjoerdsimons commented on gitlab-runner/src/job.rs:
```diff
@@ -91,6 +105,35 @@ impl Dependency<'_> {
+        &self,
+        writer: Arc<AsyncMutex<impl AsyncWrite + Unpin>>,
+    ) -> Result<(), ClientError> {
+        let strategy = FibonacciBackoff::from_millis(100).map(jitter).take(10);
```
100ms seems aggressive; 10 times without jitter is:
```
=> 100ms
=> 100ms
=> 200ms
=> 300ms
=> 500ms
=> 800ms
=> 1.3s
=> 2.1s
=> 3.4s
=> 5.5s
Total: 14.3s
```

starting from 500 would be:
```
> 500ms
=> 500ms
=> 1s
=> 1.5s
=> 2.5s
=> 4s
=> 6.5s
=> 10.5s
=> 17s
=> 27.5s
Total: 71.5s
```

Ofcourse this is a bit finger in the air as there is no way to know what he remote issue is; But a less aggressive retry might probably be better

sjoerdsimons committed:
Open-code download-retry loop

The default Action implementation as was used wants an FnMut. This then
requires locking (as well as reference count), even though the retries
are executed in purely a serial fashion.

Avoid this by simply open-coding the retry loop.

sjoerdsimons commented:
@refi64 pushed a commit on top to get rid of the arc mutex wrapping. Could you give me your opinion on it ?

refi64 commented on gitlab-runner/src/job.rs:
```diff
@@ -105,33 +102,29 @@ impl Dependency<'_> {
+                .download_artifact(self.dependency.id, &self.dependency.token, &mut writer)
+                .await;
+
+            match (r, strategy.next()) {
```
oh this is really quite slick, I hadn't considered that we could just *iterate over the strategy ourselves* by nature of it being an iterable, I guess I had the other dead-ish backoff library I used in obs-gitlab-runner stuck in my head. (might also move that over at some point, because the use of closures has—again—caused my an incredible amount of lifetime headaches.)

refi64 commented:
conflicts resolved
